File: damon/src/main/java/com/zyhang/damon/PresenterStorage.java
package com.zyhang.damon;


import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.collection.ArrayMap;

/**
 * ProjectName:Damon
 * Description:
 * Created by zyhang on 2017/4/28.22:56
 * Modify by:
 * Modify time:
 * Modify remark:
 */

enum PresenterStorage {
    INSTANCE;

    private ArrayMap<String, MvpPresenter> idToPresenter = new ArrayMap<>();
    private ArrayMap<MvpPresenter, String> presenterToId = new ArrayMap<>();

    void add(List<? extends MvpPresenter> presenters) {
        for (MvpPresenter presenter : presenters) {
            String id = presenter.getClass().getSimpleName() + "/" + System.nanoTime() + "/" + (int) (Math.random() * Integer.MAX_VALUE);
            idToPresenter.put(id, presenter);
            presenterToId.put(presenter, id);
            presenter.addOnDestroyListener(() -> idToPresenter.remove(presenterToId.remove(presenter)));
        }
    }

    @Nullable
    List<? extends MvpPresenter> getPresenter(@Nullable String[] ids) {
        if (null == ids) {
            return null;
        }
        List<MvpPresenter> presenters = null;
        for (String id : ids) {
            //noinspection unchecked
            MvpPresenter presenter = idToPresenter.get(id);
            if (presenter != null) {
                if (null == presenters) {
                    presenters = new ArrayList<>();
                }
                presenters.add(presenter);
            }
        }
        return presenters;
    }

    @Nullable
    String getId(MvpPresenter presenter) {
        return presenterToId.get(presenter);
    }
}

File: backup/src/main/kotlin/com/zyhang/damon/kotlin/PresenterStorage.kt
package com.zyhang.damon.kotlin

/**
 * ProjectName:Damon
 * Description:
 * Created by zyhang on 2017/4/28.下午5:54
 * Modify by:
 * Modify time:
 * Modify remark:
 */

internal enum class PresenterStorage {
    INSTANCE;

    val mIdToPresenter: HashMap<String, MvpPresenter<MvpView>> = HashMap()
    val mPresenterToId: HashMap<MvpPresenter<MvpView>, String> = HashMap()

    fun add(presenter: MvpPresenter<MvpView>) {
        val id: String = presenter::class.simpleName + "/" + System.nanoTime() + "/" + Math.random() * Int.MAX_VALUE
        mIdToPresenter.put(id, presenter)
        mPresenterToId.put(presenter, id)
        presenter.addOnDestroyListener(object : OnDestroyListener {
            override fun onDestroy() {
                mIdToPresenter.remove(mPresenterToId.remove(presenter))
            }
        })
    }

    @Suppress("UNCHECKED_CAST")
    fun <P> getPresenter(id: String): P = mIdToPresenter[id] as P

    fun getId(presenter: MvpPresenter<MvpView>): String = mPresenterToId[presenter]!!
}